### PR TITLE
Re-enable type checking for distributed_c10d.py

### PR DESCRIPTION
**Pull request by chipturner**
Re-enable type checking for distributed_c10d.py

Type checking for distributed_c10d.py was inadvertently turned off in issues that have accumulated since.

Note: the backwards compatibility linter does not like some of these changes.  But they were incorrect before.  This needs human verification, however.

#suppress-api-compatibility-check

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @kiukchung @lucasllc

### CODE REVIEW

**Review by wconstab — CHANGES_REQUESTED** [review]
ideep submodule was accidentally updated?

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -971,17 +971,18 @@ def _is_barrier_after_init() -> int:
...
         raise ValueError(
             "Default process group has not been initialized, "
             "please make sure to call init_process_group."
         )
+    assert GroupMember.WORLD is not None
     return GroupMember.WORLD
 
 
-def _get_default_store():
+def _get_default_store() -> Optional[Store]:
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py:985]: is this actually Optional?

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -3991,7 +3995,7 @@ def _get_group_tag(pg: ProcessGroup) -> str:
 def _get_process_group_name(pg: ProcessGroup) -> str:
     return _world.pg_names.get(pg, "None")
 
-def _get_process_group_store(pg: ProcessGroup) -> Store:
+def _get_process_group_store(pg: ProcessGroup) -> Optional[Store]:
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py:3998]: is this optional

**Comment by wconstab** [comment]
otherwise lgtm, did a semi-careful review of the annotations but skimmed over some.  Checked the .pyi more carefully

**Comment by wconstab** [comment]
ideep module still looks borked, lmk if you need help with that. ill stamp once fixed

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (chipturner) [reply torch/distributed/distributed_c10d.py:987]: good catch, the underlying map is unnecessarily `Optional[Store]`; will fix

**Comment by chipturner** [comment]
> ideep submodule was accidentally updated?

😬 wow no idea how that snuck in.  will fix.

File: torch/utils/_typing_utils.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+from typing import Optional, TypeVar
+
+# Helper to turn Optional[T] into T when we know None either isn't
+# possible or should trigger an exception.
+T = TypeVar("T")
+
+
+def not_none(obj: Optional[T]) -> T:
+    if obj is None:
+        raise RuntimeError(
```
Comment (Skylion007) [inline torch/utils/_typing_utils.py:12]:
```suggestion
        raise TypeError(
```
Shouldn't this technically raise a TypeError? or a ValueError?

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -1042,13 +1045,13 @@ def get_backend(group: Optional[ProcessGroup] = None) -> str:
...
     assert pg_store is not None
-    return pg_store[0]
+    return Backend(pg_store[0])
 
 
 _exception_logger
 @_time_logger
 def init_process_group(
-    backend: Union[str, Backend] = None,
+    backend: Optional[Backend] = None,
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py:1054]: should this be Optional[Union[str, Backend]]?

File: torch/distributed/distributed_c10d.py [hunk 5]
```diff
@@ -425,15 +427,15 @@ def __init__(self, op: Callable, tensor: torch.Tensor, dst_tensor: Optional[torc
...
+_pg_map: Dict[ProcessGroup, Tuple[str, Store]] = {}
 _pg_names: Dict[ProcessGroup, str] = {}
 _pg_group_ranks: Dict[ProcessGroup, Dict[int, int]] = {}
 # For a pg, it is a map from ProcessGroup to BackendConfig
 _pg_backend_config: Dict[ProcessGroup, str] = {}
 _group_count = 0
 _tags_to_pg: Dict[str, List[ProcessGroup]] = {}
 _pg_to_tag: Dict[ProcessGroup, str] = {}
-
+_backend: Optional[str] = None
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py:438]: should this be Optional[Union[str, Backend]] too or is this actually just str

**Review by wconstab — APPROVED** [review]
going to stamp to unblock. thanks for fixing these!

File: torch/distributed/distributed_c10d.py (hunk 4 above)
Comment (chipturner) [reply torch/distributed/distributed_c10d.py:1054]: so part of this change was to change Backend to be derived from str in addition to the metaclass magic.  this simplified some of these unions.  I actually though am surprised the type checker accepts this as I think it should be `Optional[str]`

I think @time_logger may be interfering with mypy...

Comment (chipturner) [reply torch/distributed/distributed_c10d.py:1054]: yes!  the decorators were interfering.  fixed.  (also notice `_exception_logger` isn't a decorator above... fixing that too, apparently it's always been broken)

File: torch/distributed/distributed_c10d.py [hunk 6]
```diff
@@ -149,7 +150,7 @@ def supports_complex(reduceOp: ReduceOp) -> bool:
     return reduceOp not in denyList
 
 
-class Backend:
+class Backend(str):
```
Comment (albanD) [inline torch/distributed/distributed_c10d.py:153]: Is this really a string subclass?

File: torch/distributed/distributed_c10d.py [hunk 7]
```diff
@@ -971,17 +973,18 @@ def _is_barrier_after_init() -> int:
...
 
-def _get_default_group():
+def _get_default_group() -> ProcessGroup:
     """Get the default process group created by init_process_group."""
     if not is_initialized():
         raise ValueError(
             "Default process group has not been initialized, "
             "please make sure to call init_process_group."
         )
+    assert GroupMember.WORLD is not None
```
Comment (albanD) [inline torch/distributed/distributed_c10d.py:983]: Is this any different than the not_none() used above?

File: torch/distributed/distributed_c10d.py [hunk 8]
```diff
@@ -1042,13 +1045,12 @@ def get_backend(group: Optional[ProcessGroup] = None) -> str:
         raise ValueError("Invalid process group specified")
     pg_store = _world.pg_map[pg] if pg in _world.pg_map else None
     assert pg_store is not None
-    return pg_store[0]
-
+    return Backend(pg_store[0])
 
-_exception_logger
+@_exception_logger
```
Comment (albanD) [inline torch/distributed/distributed_c10d.py:1050]: Great catch! That is definitely going to change behavior

**Review by albanD — COMMENTED** [review]
BC linter skip sounds fair given the function indeed doesn't expect that type.
I do have a couple comments to point out places where it will definitely change the behavior to make sure they're ok.

File: torch/distributed/distributed_c10d.py (hunk 6 above)
Comment (chipturner) [reply torch/distributed/distributed_c10d.py:153]: Yes... and it always has just been a str factory.  It's a very weird class.  The `__new__` method actually just returned a string always.  AFAICT there has never even been an instance of one of these classes!  We had a bunch of Union types for some reason.  But what the code seemed to really be wanting to do _is_ properly done subclassing a built-in.  So I think it's okay here, and lets us later actually potentially create real instances of these classes.

File: torch/distributed/distributed_c10d.py (hunk 7 above)
Comment (chipturner) [reply torch/distributed/distributed_c10d.py:983]: good catch, I think this predated my not_none helper.  will fix (and a handful of other similar callsites)

**Pull request closed without merging**